scripts/emacs_lua_complete.lua

```diff
@@ -0,0 +1,79 @@
+--    __emacs_lua_complete:
+-- This function finds and prints a limited number of matches on the
+-- lua globals tree, _G. See `lua-complete-string`. N.B.: Does not
+-- complete numeric keys (e.g. for arrays) or keys which begin with an
+-- underscore.
+-- Arguments:
+--   PARTS: a table array of dot-separated parts of the completion item
+--     (e.g. table.con would become {"table","con"}).  All but the last
+--     element must be valid subkeys in the _G heirarchy.
+--   LIBS: an array of tables of the form
+--       {var: varname, lib: libfile}
+--     for e.g. local varname=require("libfile"). This library will be
+--     require'd and searched alongside global variables.
+--   LOCALS: an array of strings parsed from local aloc,bloc = ... to
+--     complete as well
+--
+-- Example:
+-- __emacs_lua_complete({'table','ins'},{{var="xyz",lib="libfile"}},
+--                      {"frank","beans"},5000)
+--
+-- Written 2019/05 J.D. Smith
+function __emacs_lua_complete(parts, libs, locals, limit)
+   -- Setup a FIFO queue for breadth-first tree traversal
+   local queue = {head = 0, tail = 0}
+   function queue:push(value)
+      self.tail = self.tail+1  self[self.tail] = value
+   end
+   function queue:pop()
+      self.head = self.head+1  return self[self.head]
+   end
+
+   -- Mirror the globals table
+   local globals,cnt,limit = {}, 0, limit or 2000
+   for k,v in pairs(_G) do globals[k] = v end
+   
+   -- Add libs and locals
+   if libs then
+      for _,v in ipairs(libs) do
+	 local good,lib=pcall(require,v.lib)
+	 globals[v.var] = good and lib or {}
+      end
+   end
+   if locals then
+      for _,v in ipairs(locals) do globals[v] = {} end
+   end
+
+   -- Descend the tree to the limit of dotted parts
+   local g,pre = globals, ""
+   for i = 1,#parts do
+      if not g then break end	-- should always exist
+      if i < #parts then -- keep descending
+	 pre = (i > 1 and pre.."." or "") .. parts[i]
+	 g = g[parts[i]]
+      else  -- final part; seed the queue with any matching tables
+	 for k,v in pairs(g) do
+	    if type(k) == "string" and k:sub(1,1) ~= "_" and k:find('^'..parts[i]) then
```

Comment:
Also, If I'm reading correctly, this ignores any keys that start with underscore to avoid "special" keys. In Python at least it's common to want them shown at least in some situations, so usually it is a configuration parameter on the Emacs side. Do you think it's worth to let underscore-prefixed keys through at least if `parts[i]` starts with an underscore?

Reply:
Yes, `_key` is ignored.  My experience is in lua this is a warning for "don't go there, internal variables/methods", but we could have an optional variable to include it (passing in as another parameter to the completion lua function). Two underscores are used for all the metatable stuff, so that probably should be trimmed to avoid circularly connected trees.  This should probably be done, but I'd prefer to put it in another PR.

Reply:
Ok, I see the point in not returning them by default, but would you say that if I enter `mytable.__meta` I should receive `__metatable` as a completion candidate?

Reply:
Hmmm... I suppose so for people hacking on modules.  I guess I worry about metatable creating circularly linked trees.  But it cuts off the search so it's not the end of the world.  I could turn it back on by default and see how it works.  In my limited experience (with the `hs` module), it adds a lot of unwanted noise.

Reply:
We can keep filtering out the special keys in the breadth search, this only applies to when there is a prefix to filter that indicates that the user wants to complete an underscore-prefixed key.

Reply:
I see, so if they explicitly complete on "_xxx[M-Tab]".  That seems reasonable. Just pushed a commit, and took the opportunity to remove the regex based match, since it could be fooled by special characters in the completion fragment.  So it doesn't offer _ keys by default, but if you start with _, it will give you everything.

Reply:
Let me know if this works as you'd expect for _XXX keys.

Reply:
Yeah, looks fine.